# Tanimoto coefficient のPythonでの実装

NumPyパッケージのインポート

In [81]:
import numpy as np

seed=1としているが、これは、乱数が生成される値を固定するためである  
いろいろ試したい方は、以下のコードをコメントアウトして（コードの前に"<b> # </b>"を書いて）実行してください

In [82]:
np.random.seed(seed=1)

正解のラベルと予測値をそれぞれ $F_{GT}$, $F_{PRED}$ とする  
今回は乱数で0または1値をもつ $ (3 \times 3) $ の行列を生成した

In [83]:
F_GT_0 = np.random.randint(0, 2, (3, 3))
F_PRED_0 = np.random.randint(0, 2, (3, 3))

In [86]:
print(F_GT_0)
print(F_PRED_0)

[[1 1 0]
 [0 1 1]
 [1 1 1]]
[[0 0 1]
 [0 1 1]
 [0 0 1]]


Expected Answer:  
$ F_{GT} $ :
$$
\begin{pmatrix}
1 & 1 & 0 \\
0 & 1 & 1 \\
1 & 1 & 1
\end{pmatrix}
$$
$ F_{PRED} $ :
$$
\begin{pmatrix}
0 & 0 & 1 \\
0 & 1 & 1 \\
0 & 0 & 1
\end{pmatrix}
$$

In [87]:
print('行列の形 : ', F_GT_0.shape)
print('行列の形 : ', F_PRED_0.shape)

行列の形 :  (3, 3)
行列の形 :  (3, 3)


表形式でまとめると、以下のようになる

| | GT_True | GT_False | Sum |
| - | - | - | - |
| **PRED_True** | 3 | 1 | 4 |
|**PRED_False** | 4 | 1 | 5 |
| **Sum** | 7 | 2 | 9|

---
## Tanimoto coefficient
U-Netの損失関数として定義されているTanimoto coefficientは以下のような式である

>$$
Loss = 1 \times T_C(F_{GT}, F_{PRED})  \\
T_C(F_{GT}, F_{PRED}) = {F_{GT} \cdot F_{PRED}} / {(|F_{GT}|^2 + |F_{PRED}|^2 - F_{GT} \cdot F_{PRED})}
$$

下記がTanimoto coefficientをPythonで実装したものである

In [67]:
def Tanimoto_coef(gt, pred):
    coef = np.dot(gt, pred)/(np.dot(gt, gt)+np.dot(pred, pred)-np.dot(gt, pred))
    return coef

先ほど生成した$ F_{GT} $, $ F_{PRED} $のTanimoto coefficientを計算するために、１次元配列に形を整える

In [69]:
F_GT = F_GT_0.reshape([-1])
F_PRED = F_PRED_0.reshape([-1])
print(F_GT)
print(F_PRED)

[1 1 0 0 1 1 1 1 1]
[0 0 1 0 1 1 0 0 1]


生成した１次元配列をTanimoto coefficient関数に代入して計算

In [70]:
T_C = Tanimoto_coef(F_GT, F_PRED)
T_C

0.375

なお、Tanimoto coefficientを手計算すると 3/(7+4-3) = 3/(11-3) = 3/8 = 0.375となり、上の値と一致する  
また、すべて間違った場合と正解した場合は以下のようになり、確かに **[0, 1]** の範囲である

In [72]:
TMP_0 = np.zeros(9).astype(int)
TMP_1 = np.ones(9).astype(int)
print(TMP_0)
print(TMP_1)

[0 0 0 0 0 0 0 0 0]
[1 1 1 1 1 1 1 1 1]


In [73]:
# 正解はすべて0、予測はすべて1
print(Tanimoto_coef(TMP_0, TMP_1))

0.0


In [74]:
# 正解はすべて1、予測はすべて0
print(Tanimoto_coef(TMP_1, TMP_0))

0.0


In [76]:
# 正解も予測も1
print(Tanimoto_coef(TMP_1, TMP_1))

1.0


なお、すべての正解のラベルと予測値が0の場合、Warningとなる(関数はnanを返す)

In [77]:
# 正解も予測も0
print(Tanimoto_coef(TMP_0, TMP_0))

nan


C:\Users\yuuki\anaconda3\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: invalid value encountered in long_scalars
  
